text/2019-11-21-key-visualizer.md

```diff
@@ -0,0 +1,295 @@
+first use to troubleshoot performance problems. After investigation, we realized
+that it has good features that also suits `TiKV`, thus we added some of those
+features to our design.
+
+## Detailed Design
+
+### Features
+
+`KeyVisualizer` is enabled by default. It collects various statistics information
+in regular intervals, and renders multiple heatmaps in among the last hour/24 hours/week/month.
+
+#### Frontend Features
+
+The heatmap is a two-dimensional color chart where:
+
+- X-axis: time axis
+- Y-axis: key axis
+- Color: value
+
+`KeyVisualizer` provides several different heatmaps for:
+
+- Read traffic
+- Read Ops (long-term plan)
+- Write traffic
+- Write Ops (long-term plan)
+- Read latency (long-term plan)
+- Write latency (long-term plan)
+- CPU usage (long-term plan)
+- Memory usage (long-term plan)
+- Disk read and write (long-term plan)
+- Number of client connections (long-term plan)
+
+Interactions:
+
+- Display label axis showing DB name, table name, and row id of the current row
+- When hover, display a popup containing the key range,
+    time range and all known statistics of the cell under the cursor
+- Warn abnormal area (long-term plan):
+- Traffic exceeds threshold
+  - Latency or wait time is too long
+  - Brush select or click on a label to zoom
+
+### Backend Features
+
+- Extend `PD` API to provide necessary data for frontend
+- Persist statistics data
+- ​​Support decoding various key encoding (default `TiDB` encoding)
+
+### `PD` Frontend Implementation
+
+#### Retrieve Data
+
+The frontend gets heatmap data via the new PD API.
+
+The API interface is defined as:
+
+URL:  `/pd/apis/keyvisual/v1/heatmaps?`
+
+Request parameters:
+
+- `type`: indicating which heatmap to show
+  - `write_bytes`
+  - `read_bytes`
+  - `write_keys`
+  - `read_keys`
+  - `startkey`: the start of the heatmap key axis
+  - `endkey`:  the end of the heatmap key axis
+  - `starttime`: the start of the heatmap time axis (UNIX timestamp in second)
+  - `endtime`: the end of the time axis of the heatmap (UNIX timestamp in second)
+
+Respond data:
+
+```typescript
+export type KeyAxisEntry = {
+  key: string
+  labels: string[]
+}
+
+export type HeatmapData = {
+  timeAxis: number[]
+  keyAxis: KeyAxisEntry[]
+  values: number[][]
+}
+```
+
+#### Render the Heatmap
+
+Render the heatmap once data retrieved.
+
+- Use `D3.js` to support the time axis, the key axis, and the Label axis.
+- Use `OffscreenCanvas` for background rendering,
+    when each pixel is corresponding to a data cell.
+    At least 1000 x 1000 cells can be displayed in one page.
+
+POC work:
+
+> workload:
+> sysbench oltp_update_non_index -table-size=4
+
+![key-visualizer-poc](../media/key-visualizer-poc.png)
+
+### PD Backend Implementation
+
+#### Collect Region Statistics
+
+Retrieve statistics data among all regions per miniute from `PD Server`.
+
+Unresolved problems:
+
+- The number of regions can reach up to millions.
```

Comment:
Perhaps the regions can report the metrics, as opposed to needing to retrieve them all. This way regions can report only on change.